solver/include/solver/solver.hpp

```diff
@@ -4,15 +4,308 @@
 
 #pragma once
 
-#include <iostream>
-#include <vector>
+#include <cmath>
+#include <iterator>
+#include <limits>
 
 #include "method.hpp"
 #include "time_span.hpp"
 
 namespace ode
 {
 
+    /**
+     * @brief store \f$(t^n, u^n, \Delta t^n)\f$
+     *
+     * @tparam value_t type of time
+     * @tparam state_t type of solution \f$u^n\f$
+     */
+    template <typename value_t, typename state_t>
+    struct current_solution
+    {
+        value_t time;
+        state_t state;
+        value_t time_step;
+
+        current_solution( value_t tn, state_t un, value_t dt )
+            : time( tn )
+            , state( un )
+            , time_step( dt )
+        {
+        }
+    };
+
+    template <typename value_t, typename state_t, typename method_t, typename problem_t>
+    struct time_iterator
+    {
+        using difference_type   = value_t;
+        using value_type        = current_solution<value_t, state_t>;
+        using pointer           = value_type*;
+        using reference         = value_type&;
+        using const_pointer     = value_type const*;
+        using const_reference   = value_type const&;
+        using iterator_category = std::output_iterator_tag;
+
+        value_type sol;
+        method_t meth;
+        problem_t& pb;
+        ponio::time_span<value_t> t_span;
+        typename ponio::time_span<value_t>::iterator it_next_time;
+        value_t final_time; // should be a ponio::time_span;
+        static constexpr value_t sentinel = std::numeric_limits<value_t>::max();
+
+        // time_iterator ( problem_t & pb_, method_t meth_, state_t const& u0, ponio::time_span<value_t> && times, value_t dt )
+        // : sol(times.front(), u0, dt)
+        // , meth(meth_)
+        // , pb(pb_)
+        // , t_span(times)
+        // , it_next_time(++times.begin())
+        // {
+        // }
+
+        time_iterator( problem_t& pb_, method_t meth_, state_t const& u0, ponio::time_span<value_t> const& t_span_, value_t dt )
+            : sol( ( t_span_.front() == t_span_.back() ) ? sentinel : t_span_.front(), u0, dt )
+            , meth( meth_ )
+            , pb( pb_ )
+            , t_span( t_span_ )
+            , it_next_time( std::begin( t_span ) )
+            , final_time( t_span.back() )
+        {
+        }
+
+        time_iterator( time_iterator const& rhs )
+            : sol( rhs.sol )
+            , meth( rhs.meth )
+            , pb( rhs.pb )
+            , t_span( rhs.t_span )
+            , it_next_time( std::begin( t_span ) + ( rhs.it_next_time - std::begin( rhs.t_span ) ) )
```

Comment:
Why is not just a copy of `rhs.it.next_time`?

Reply:
Because I want an iterator on `t_span` not on `rhs.t_span`. So I need to compute the index of `rhs.it_next_time` and shift `it_next_time` to the correct value (on `t_span`).

Reply:
So maybe you can use `std::distance`?

Reply:
I can't because `std::begin` return a `const_iterator` and `rhs.it_next_time` is an `iterator`, so I can use `std::ranges::distance` (pretty much the same function but with different input types).